# Module 06 Lab - Regression and Classification Models**Objective:** To understand the difference between regression and classification and to build your first linear models for both tasks.**In this lab, you will write the code to train and evaluate the models.**

## Part 1: Regression vs. ClassificationThis is the most fundamental distinction between types of supervised learning problems.*   **Regression:** The goal is to predict a **continuous numerical value**.     *   *Examples:* Predicting the price of a house, the temperature tomorrow, or the stock price.*   **Classification:** The goal is to predict a **discrete category or class label**.    *   *Examples:* Predicting if an email is spam or not spam, if a flower is a setosa, versicolor, or virginica, or if a customer will churn or not.**In this lab, we will tackle one of each.**

## Part 2: Linear Regression**Concept:** Linear Regression is used to predict a continuous value. It works by finding the best-fitting straight line through the data points. The model learns a "slope" (coefficient) for each feature and an "intercept".*   **Problem:** We will predict the `Fare` of a Titanic passenger based on their `Age` and `Pclass`.

In [29]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

# data
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')
df.dropna(subset=['Age'], inplace=True)


features = ['Age', 'Pclass']
target_reg = 'Fare'
X_reg = df[features]
y_reg = df[target_reg]

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(X_reg, y_reg, test_size=0.2, random_state=42)

### Task 1: Train and Evaluate a Linear Regression Model**Your Task:**1.  Create an instance of the `LinearRegression` model.2.  Train the model using the training data (`X_train_reg`, `y_train_reg`).3.  Make predictions on the test data.4.  Evaluate the model using `mean_squared_error`. This metric tells us the average of the squared differences between the predicted and actual values.

In [30]:
lin_reg = LinearRegression()
lin_reg.fit(X_train_reg, y_train_reg)
y_pred_reg = lin_reg.predict(X_test_reg)
print(mean_squared_error(y_test_reg, y_pred_reg))


3364.922912856981


The error occurred because several import statements were written together on a single line, which caused Python to raise a SyntaxError. Import statements must either be placed on separate lines or separated with semicolons. Gemini fixed the issue by moving each import statement onto its own line so the code could run correctly. After correcting the imports, the data loading, preprocessing, and train‑test split for the regression task executed worked.

## Part 3: Logistic Regression**Concept:** Despite its name, Logistic Regression is a **classification** algorithm. It works by calculating the probability that a given input belongs to a certain class. It's one of the most widely used and interpretable classification models.*   **Problem:** We will predict whether a passenger `Survived` based on their `Age`, `Pclass`, and `Sex`.

In [31]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Prepare data for classification
# We need to encode 'Sex' column
df['Sex'] = df['Sex'].map({'male': 0, 'female': 1})

# Define features and target
features_cls = ['Age', 'Pclass', 'Sex']
target_cls = 'Survived'
X_cls = df[features_cls]
y_cls = df[target_cls]

# Split the data
X_train_cls, X_test_cls, y_train_cls, y_test_cls = train_test_split(X_cls, y_cls, test_size=0.2, random_state=42)

### Task 2: Train and Evaluate a Logistic Regression Model**Your Task:**1.  Create an instance of the `LogisticRegression` model.2.  Train the model using the classification training data.3.  Make predictions on the test data.4.  Evaluate the model using `accuracy_score`.

In [32]:
# 1. Create the model instance
log_model = LogisticRegression(max_iter=1000)

# 2. Train the model
log_model.fit(X_train_cls, y_train_cls)

# 3.predict
y_pred_cls = log_model.predict(X_test_cls)

accuracy = accuracy_score(y_test_cls, y_pred_cls)
print("Accuracy:", accuracy)


Accuracy: 0.7482517482517482


In this step, I (Trevor) trained a Logistic Regression model to classify whether a passenger survived the Titanic. I used Age, Pclass as the input features and split the data into training and testing sets. Then predictions were made and calculated the accuracy score to see how well the model performed.

In [33]:
#try 1

from sklearn.preprocessing import StandardScaler

# Initialize the StandardScaler
scaler = StandardScaler()

# Fit the scaler on the training data and transform both training and test data
X_train_cls_scaled = scaler.fit_transform(X_train_cls)
X_test_cls_scaled = scaler.transform(X_test_cls)

# Retrain the Logistic Regression model with scaled data
log_model_scaled = LogisticRegression(max_iter=1000)
log_model_scaled.fit(X_train_cls_scaled, y_train_cls)

# Make predictions on the scaled test data
y_pred_cls_scaled = log_model_scaled.predict(X_test_cls_scaled)

# Evaluate the model with the new predictions
accuracy_scaled = accuracy_score(y_test_cls, y_pred_cls_scaled)
print(f"--- Try 1: Feature Scaling ---")
print(f"Initial Logistic Regression Accuracy: {accuracy:.4f}")
print(f"Logistic Regression Accuracy (after scaling): {accuracy_scaled:.4f}")



#try 2

from sklearn.ensemble import RandomForestClassifier

# Initialize the Random Forest Classifier
# Using 'random_state' for reproducibility
# You can tune 'n_estimators' and 'max_depth' further
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

# Train the model on the scaled training data
rf_model.fit(X_train_cls_scaled, y_train_cls)

# Make predictions on the scaled test data
y_pred_rf = rf_model.predict(X_test_cls_scaled)

# Evaluate the Random Forest model's accuracy
accuracy_rf = accuracy_score(y_test_cls, y_pred_rf)
print(f"\n--- Try 2: Random Forest Classifier ---")
print(f"Random Forest Classifier Accuracy: {accuracy_rf:.4f}")


#try 3


from sklearn.model_selection import GridSearchCV

# Define the parameter grid for Logistic Regression
param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'solver': ['liblinear', 'lbfgs', 'saga']
}

# Initialize GridSearchCV
# We use the scaled training data for tuning as it can sometimes help solvers converge faster
grid_search = GridSearchCV(LogisticRegression(max_iter=1000), param_grid, cv=5, scoring='accuracy', n_jobs=-1)

# Fit GridSearchCV to the scaled training data
grid_search.fit(X_train_cls_scaled, y_train_cls)

# Print the best parameters and best score
print(f"\n--- Try 3: Hyperparameter Tuning (Logistic Regression) ---")
print(f"Best Parameters: {grid_search.best_params_}")
print(f"Best Cross-validation Accuracy: {grid_search.best_score_:.4f}")

# Get the best model from the grid search
best_log_model = grid_search.best_estimator_

# Make predictions on the scaled test data with the best model
y_pred_best = best_log_model.predict(X_test_cls_scaled)

# Evaluate the best model's accuracy on the test set
accuracy_best = accuracy_score(y_test_cls, y_pred_best)
print(f"Test Accuracy with Best Tuned Model: {accuracy_best:.4f}")

--- Try 1: Feature Scaling ---
Initial Logistic Regression Accuracy: 0.7483
Logistic Regression Accuracy (after scaling): 0.7413

--- Try 2: Random Forest Classifier ---
Random Forest Classifier Accuracy: 0.7483

--- Try 3: Hyperparameter Tuning (Logistic Regression) ---
Best Parameters: {'C': 1, 'solver': 'liblinear'}
Best Cross-validation Accuracy: 0.8127
Test Accuracy with Best Tuned Model: 0.7413


AI Disclaimer: The test cases, were generated with the assistance of Gemini and some changes by Ranveer.

Prompt: "Implement three Python code blocks to attempt improving Logistic Regression accuracy without feature engineering. First, apply StandardScaler, retrain Logistic Regression, and print original and scaled accuracies. Second, train a RandomForestClassifier on scaled data and print its accuracy. Finally third, use GridSearchCV to tune Logistic Regression (C and solver) on scaled data and print best parameters, cross-validation, and test accuracy. Make sure all results are printed."

### Summary of Model Improvement Attempts Above

Try 1 (Feature Scaling): We applied StandardScaler to numerical features, retraining the Logistic Regression model. This resulted in a slight decrease in accuracy from 0.7483 to 0.7413.

Try 2 (Random Forest Classifier): We then tried a more complex RandomForestClassifier on the scaled data. This model achieved an accuracy of 0.7483, which was identical to the original Logistic Regression baseline.

Try 3 (Hyperparameter Tuning): Finally, GridSearchCV was used to tune the C parameter and solver for the Logistic Regression model on the scaled data. Although cross-validation accuracy reached 0.8127, the test accuracy remained 0.7413.

The almost same test accuracy scores across these trials shows that with the current features, the models have reached a limit, even with different hyperparameter tuning, unless new features are engineered we will be capped around this score.


## 📝 Knowledge Check
Instructions: Answer the following questions in this markdown cell.
1. In your own words, what is the key difference between a regression problem and a classification problem?

A regression problem predicts a continuous numerical value (like price, temperature,  fare, 7.05, 10.50, 70.83, or any other number.).
A classification problem predicts a category or class label (like survived/not survived, spam/not spam).

2. The LinearRegression model has an attribute called .coef_. After you train the model, print lr_model.coef_. What do these numbers represent?

lr_model.coef_ contains the learned weights (slopes) for each feature. Each number shows how much the predicted Fare changes when that feature increases by one unit, holding other features constant.

3. Why did we use mean_squared_error to evaluate the regression model but accuracy_score for the classification model? Why wouldn't accuracy be a good metric for the fare prediction task?

Regression needs a metric that measures how far predictions are from the true numeric values, so we use mean_squared_error.
Classification needs a metric that measures how many labels were predicted correctly, so we use accuracy_score.
Accuracy would be bad for fare prediction because fare is a continuous number,  the model would almost never guess the exact fare, so accuracy would always be near zero.
